File: Assets/Bundles/Client/Resources/TS/Controllers/Global/ProximityPrompt/ProximityPrompt.lua
-- Compiled with unity-ts v2.1.0-75
local ClientSignals = require("Client/TS/ClientSignals").ClientSignals
local GameObjectBridge = require("Shared/TS/GameObjectBridge").GameObjectBridge
local Signal = require("Shared/TS/Util/Signal").Signal
-- * Proximity Prompt.
local ProximityPrompt
do
	ProximityPrompt = setmetatable({}, {
		__tostring = function()
			return "ProximityPrompt"
		end,
	})
	ProximityPrompt.__index = ProximityPrompt
	function ProximityPrompt.new(...)
		local self = setmetatable({}, ProximityPrompt)
		return self:constructor(...) or self
	end
	function ProximityPrompt:constructor(promptData)
		self.OnActivated = Signal.new()
		self.promptPrefab = AssetBridge:LoadAsset("Client/Resources/Prefabs/ProximityPrompt.prefab")
		local _original = ProximityPrompt.idCounter
		ProximityPrompt.idCounter += 1
		self.id = tostring(_original)
		self.data = promptData
		self:CreatePrompt()
	end
	function ProximityPrompt:CreatePrompt()
		self.promptGameObject = GameObjectBridge:InstantiateAt(self.promptPrefab, self.data.promptPosition, Quaternion.identity)
		-- Prompt starts inactive.
		self.promptGameObject:SetActive(false)
		-- Set activation key, action, and object text.
		local keyCode = self.promptGameObject.transform:Find("Canvas/Background/KeyCodeBackground/KeyCode"):GetComponent("TextMeshProUGUI")
		local textWrapper = self.promptGameObject.transform:Find("Canvas/Background/TextWrapper")
		local bottomText = textWrapper:FindChild("BottomText"):GetComponent("TextMeshProUGUI")
		local topText = textWrapper:FindChild("TopText"):GetComponent("TextMeshProUGUI")
		keyCode.text = self.data.activationKeyString
		bottomText.text = self.data.bottomText
		topText.text = self.data.topText
		-- Notify `ProximityPromptController` that prompt was created.
		ClientSignals.ProximityPromptCreated:Fire({
			prompt = self,
		})
	end
	function ProximityPrompt:ActivatePrompt()
		self.OnActivated:Fire()
	end
	ProximityPrompt.idCounter = 0
end
return {
	ProximityPrompt = ProximityPrompt,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Global/ProximityPrompt/ProximityPrompt.ts
import { Dependency } from "@easy-games/flamework-core";
import { GameObjectBridge } from "Shared/GameObjectBridge";
import { Keyboard } from "Shared/UserInput";
import { Bin } from "Shared/Util/Bin";
import { Signal } from "Shared/Util/Signal";
import { ProximityPromptController } from "./ProximityPromptController";

export interface ProximityPromptData {
	/** Proximity prompt position. */
	promptPosition: Vector3;
	/** Key that activates proximity prompt. */
	activationKey: KeyCode;
	/** Activation key string that displays on prompt. */
	activationKeyString: string;
	/** How close local player must be to activate proximity prompt. */
	activationRange: number;
	/** Proximity prompt top text. */
	topText: string;
	/** Proximity prompt bottom text. */
	bottomText: string;
}

/** Proximity Prompt. */
export class ProximityPrompt {
	/** Global, incrementing id counter. */
	public static idCounter = 0;
	/** Proximity prompt prefab. */
	private promptPrefab: Object;
	/** Proximity prompt id. */
	public id: string;
	/** Proximity prompt data. */
	public data: ProximityPromptData;
	/** Proximity prompt gameobject. */
	public promptGameObject: GameObject | undefined;
	/** On activated signal. */
	public OnActivated = new Signal<void>();
	/** On activated signal. */
	public OnRequestActivated = new Signal<void>();

	private canActivate = false;
	private readonly canActivateBin = new Bin();

	constructor(promptData: ProximityPromptData) {
		this.promptPrefab = AssetBridge.LoadAsset("Client/Resources/Prefabs/ProximityPrompt.prefab");
		this.id = tostring(ProximityPrompt.idCounter++);
		this.data = promptData;
		this.CreatePrompt();
	}

	/** Creates prompt from prompt data. */
	private CreatePrompt(): void {
		this.promptGameObject = GameObjectBridge.InstantiateAt(
			this.promptPrefab,
			this.data.promptPosition,
			Quaternion.identity,
		);
		/* Prompt starts inactive. */
		this.promptGameObject.SetActive(false);
		/* Set activation key, action, and object text. */
		const keyCode = this.promptGameObject.transform
			.Find("Canvas/Background/KeyCodeBackground/KeyCode")!
			.GetComponent<TextMeshProUGUI>();
		const textWrapper = this.promptGameObject.transform.Find("Canvas/Background/TextWrapper")!;
		const bottomText = textWrapper.FindChild("BottomText")!.GetComponent<TextMeshProUGUI>();
		const topText = textWrapper.FindChild("TopText")!.GetComponent<TextMeshProUGUI>();
		keyCode.text = this.data.activationKeyString;
		bottomText.text = this.data.bottomText;
		topText.text = this.data.topText;
		Dependency<ProximityPromptController>().RegisterProximityPrompt(this);
	}

	public SetCanActivate(canActivate: boolean) {
		if (this.canActivate === canActivate) return;
		this.canActivate = true;
		if (canActivate) {
			const keyboard = this.canActivateBin.Add(new Keyboard());
			keyboard.OnKeyDown(this.data.activationKey, (event) => {
				this.OnRequestActivated.Fire();
			});
		} else {
			this.canActivateBin.Clean();
		}
	}

	/** Called when prompt activates. */
	public ActivatePrompt(): void {
		this.OnActivated.Fire();
	}
}
